File: part-3/imlib.py
"""
4190.308 Computer Architecture                                                          Spring 2023

Image Library

This library provides functions for reading and writing images in various formats. In particular, 
the library can read and write images in a custom RAW format.

@author:
    Bernhard Egger <bernhard@csap.snu.ac.kr>

@changes:
    2023/02/14 Bernhard Egger created

@license:
    Copyright (c) 2023, Computer Systems and Platforms Laboratory, SNU
    All rights reserved.

    Redistribution and use in source and binary forms, with or without modification, are permitted
    provided that the following conditions are met:

    - Redistributions of source code must retain the above copyright notice, this list of condi-
      tions and the following disclaimer.
    - Redistributions in binary form must reproduce the above copyright notice, this list of condi-
      tions and the following disclaimer in the documentation and/or other materials provided with
      the distribution.

    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR
    IMPLIED WARRANTIES, INCLUDING,  BUT NOT LIMITED TO,  THE IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
    CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR CONSE-
    QUENTIAL DAMAGES  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
    LOSS OF USE,  DATA, OR PROFITS; OR BUSINESS INTERRUPTION)  HOWEVER CAUSED AND ON ANY THEORY OF
    LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
    ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
    DAMAGE.
"""


"""
CSAP RAW image format specification

The CSAP RAW image format contains raw uncompressed pixel data. The data is 


  MAGIC   4 characters 'C', 'S', 'A', 'P'.
  FORMAT  integer, big endian. Valid values: BGR_FORMAT (3 channels) or BGRA_FORMAT (4 channels).
  Height  integer, little endian. Specifices the height of the image.
  Width   integer, little endian. Specifices the width of the image.
  Data    image data stored as multi-dimensional array, row-major.
"""

import cv2
import numpy as np
import struct

# Constants
MAGIC = 0x43534150 # 'CSAP'
BGR_FORMAT  = 0x4247522d  # 'BGR-'
BGRA_FORMAT = 0x42475241  # 'BGRA'


def read_image(filename):
    """
    Reads a (compressed) image file and returns its pixel data, height, width, and # channels.

    Args:
        filename: A string specifying the filename of the image file.

    Returns:
        A tuple containing the following elements:
        - A multi-level list encoding the image data. Each inner list contains the pixel 
          values for one row of the image.
        - The height of the image (an integer).
        - The width of the image (an integer).
        - The number of channels in the image (either 3 or 4).
    """
    try:
        img = cv2.imread(filename, cv2.IMREAD_UNCHANGED)
        if img is None:
            raise Exception(f"Error: Could not read image file '{filename}'")
    except cv2.error as e:
        raise Exception(f"Error: {e}")

    height, width, num_channels = img.shape[:3]

    # Make sure we have three (BGR) or four (BGRA) channels
    if num_channels not in [3, 4]:
        raise ValueError('Invalid image. Only three or four channel images are supported.')

    # Convert to multi-level list
    image = [[list(img[h, w])[:num_channels] for w in range(width)] for h in range(height)]

    return image, height, width, num_channels



def read_raw_image(filename):
    """
    Reads a RAW image file and returns its pixel data, height, width, and number of channels.

    Args:
        filename: A string specifying the filename of the RAW image file.

    Returns:
        A tuple containing the following elements:
        - A multi-level list encoding the image data. Each inner list contains the pixel 
          values for one row of the image.
        - The height of the image (an integer).
        - The width of the image (an integer).
        - The number of channels in the image (either 3 or 4).
    """
    with open(filename, 'rb') as f:
        # Read the magic number
        magic = struct.unpack('>I', f.read(4))[0]
        if magic != MAGIC:
            raise ValueError('Invalid magic number')

        # Read the data format
        format_code = struct.unpack('>I', f.read(4))[0]
        if format_code == BGR_FORMAT:
            num_channels = 3
        elif format_code == BGRA_FORMAT:
            num_channels = 4
        else:
            raise ValueError('Invalid data format')

        # Read the height and width
        height = struct.unpack('<i', f.read(4))[0]
        width = struct.unpack('<i', f.read(4))[0]

        # Read the pixel data
        expected_bytes = height * width * num_channels
        data = f.read(expected_bytes)
        if len(data) != expected_bytes:
            raise ValueError(f"Tried to read {expected_bytes} bytes, but got only {len(data)}")

    # Convert the pixel data to a 3D array
    image = [[[0 for c in range(num_channels)] for w in range(width)] for h in range(height)]
    for h in range(height):
        for w in range(width):
            for c in range(num_channels):
                index = (h * width + w) * num_channels + c
                image[h][w][c] = data[index]

    return image, height, width, num_channels


def write_image(filename, image, height, width, num_channels):
    """
    Saves an image to disk.

    Args:
        filename:     A string specifying the filename of the RAW image file.
        image:        A multi-level list encoding the image data. Each inner list contains the 
                      pixel values for one row of the image.
        height:       The height of the image (an integer).
        width:        The width of the image (an integer).
        num_channels: The number of channels in the image (either 3 or 4).

    Raises:
        ValueError:   If the image data doesn't match the height, width, and number of channels
        IOError:      If there was a problem writing the output file
    """
    # Only 3 and 4 channels are supported
    if num_channels not in [3, 4]:
        raise ValueError('Invalid data format')

    # Make sure dimensions match
    if len(image) != height or len(image[0]) != width or len(image[0][0]) != num_channels:
        raise ValueError('Invalid data shape')

    # Flatten image data into a byte array
    data = bytearray()
    for h in range(height):
        for w in range(width):
            for c in range(num_channels):
                data.append(image[h][w][c])

    # Save the image using OpenCV's imwrite function
    try:
        img = np.frombuffer(data, dtype=np.uint8).reshape(height, width, -1)
        cv2.imwrite(filename, img)
    except Exception as e:
        raise IOError(f"Error writing image to file: {e}")



def write_raw_image(filename, image, height, width, num_channels):
    """
    Saves an image in RAW image file format.

    Args:
        filename:     A string specifying the filename of the RAW image file.
        image:        A multi-level list encoding the image data. Each inner list contains the 
                      pixel values for one row of the image.
        height:       The height of the image (an integer).
        width:        The width of the image (an integer).
        num_channels: The number of channels in the image (either 3 or 4).

    Raises:
        ValueError:   If the image data doesn't match the height, width, and number of channels
        IOError:      If there was a problem writing the output file
    """
    # Only 3 and 4 channels are supported
    if num_channels not in [3, 4]:
        raise ValueError('Invalid data format')

    # Make sure dimensions match
    if len(image) != height or len(image[0]) != width or len(image[0][0]) != num_channels:
        raise ValueError('Invalid data shape')

    # Flatten image data into a byte array
    data = bytearray()
    for h in range(height):
        for w in range(width):
            for c in range(num_channels):
                data.append(image[h][w][c])

    # Write data to file
    with open(filename, 'wb') as f:
        # Write the magic number
        f.write(struct.pack('>I', MAGIC))

        # Write the data format
        f.write(struct.pack('>I', (BGR_FORMAT if num_channels == 3 else BGRA_FORMAT)))

        # Write the height and width
        f.write(struct.pack('<i', height))
        f.write(struct.pack('<i', width))

        # Write the pixel data
        f.write(data)

File: part-2/imlib.c
//-------------------------------------------------------------------------------------------------
// 4190.308 Computer Architecture                                                       Spring 2023
//
/// @file
/// @brief Image Library
///        This library provides functions for reading and writing images in various formats. 
///        In particular, the library can read and write images in a custom RAW format.
///
/// @author Bernhard Egger <bernhard@csap.snu.ac.kr>
/// @section changelog Change Log
/// 2023/02/14 Bernhard Egger created
///
/// @section license_section License
/// Copyright (c) 2023, Computer Systems and Platforms Laboratory, SNU
/// All rights reserved.
///
/// Redistribution and use in source and binary forms, with or without modification, are permitted
/// provided that the following conditions are met:
///
/// - Redistributions of source code must retain the above copyright notice, this list of condi-
///   tions and the following disclaimer.
/// - Redistributions in binary form must reproduce the above copyright notice, this list of condi-
///   tions and the following disclaimer in the documentation and/or other materials provided with
///   the distribution.
///
/// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR
/// IMPLIED WARRANTIES, INCLUDING,  BUT NOT LIMITED TO,  THE IMPLIED WARRANTIES OF MERCHANTABILITY
/// AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
/// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR CONSE-
/// QUENTIAL DAMAGES  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
/// LOSS OF USE,  DATA, OR PROFITS; OR BUSINESS INTERRUPTION)  HOWEVER CAUSED AND ON ANY THEORY OF
/// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
/// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
/// DAMAGE.
//-------------------------------------------------------------------------------------------------

#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "imlib.h"


static uint8 MAGIC[4] = { 'C', 'S', 'A', 'P' };
static uint8 BGR_FORMAT[4] = { 'B', 'G', 'R', '-' };
static uint8 BGRA_FORMAT[4] = { 'B', 'G', 'R', 'A' };


void panic(char *message, int errorno)
{
  char *error = NULL;
  if (errorno !=0) error = strerror(errorno);

  fprintf(stderr, "%s%s%s\n",
          message ? message : "",
          message && error ? ": ":"",
          error ? error : "");

  exit(EXIT_FAILURE);
}


struct Image read_raw_image(char *filename)
{
  FILE *f;
  struct Image img = { NULL, -1, -1, -1 };

  // Open file
  if ((f = fopen(filename, "rb")) == NULL) panic("Cannot open file", errno);

  // Read magic number
  uint8 magic[4];
  if (fread(&magic, sizeof(magic), 1, f) < 1) panic("Cannot read magic", errno);
  if (*(int*)magic != *(int*)MAGIC) {
    char msg[64];
    snprintf(msg, sizeof(msg), "Invalid magic number: %08x (expected %08x).\n", 
             *(int*)magic, *(int*)MAGIC);
    panic(msg, 0);
  }

  // Read data format
  uint8 format[4];
  if (fread(&format, sizeof(format), 1, f) < 1) panic("Cannot read image format", errno);
  if (*(int*)format == *(int*)BGR_FORMAT) {
    img.channels = 3;
  } else if (*(int*)format == *(int*)BGRA_FORMAT) {
    img.channels = 4;
  } else {
    char msg[64];
    snprintf(msg, sizeof(msg), "Invalid data format: %08x.\n", *(int*)format);
    panic(msg, 0);
  }

  // Read height and width
  uint8 h[4], w[4];
  if (fread(h, sizeof(h), 1, f) < 1) panic("Cannot read image height", errno);
  if (fread(w, sizeof(w), 1, f) < 1) panic("Cannot read image width", errno);
  img.height = h[3] << 24 | h[2] << 16 | h[1] << 8 | h[0];
  img.width  = w[3] << 24 | w[2] << 16 | w[1] << 8 | w[0];

  // Allocate memory for image data
  int img_size = img.height * img.width * img.channels;
  if ((img.data = malloc(sizeof(uint8)*img_size)) == NULL) {
    panic("Failed to allocate memory for image", errno);
  }

  // Read pixel data
  if (fread(img.data, sizeof(uint8), img_size, f) < img_size) {
    panic("Cannot read image data", errno);
  }

  // Clean up and return
  fclose(f);

  return img;
}


void write_raw_image(char* filename, struct Image img)
{
  FILE *f;

  // Run a few checks
  if (img.data == NULL) panic("No image data.", 0);

  // Only 3 and 4 channels are supported
  if ((img.channels < 3) || (4 < img.channels)) panic("Invalid data format.", 0);

  // Write data to file
  if ((f = fopen(filename, "wb")) == NULL) panic("Cannot open file", errno);

  // Write magic number and data format (big endian)
  if (fwrite(MAGIC, sizeof(MAGIC), 1, f) < 1) panic("Cannot write magic to file", errno);
  uint8 *format = (img.channels == 3 ? BGR_FORMAT : BGRA_FORMAT);
  if (fwrite(format, sizeof(BGR_FORMAT), 1, f) < 1) panic("Cannot write format to file", errno);

  // Write height and width (little endian)
  uint8 h[4] = {img.height, img.height>>8, img.height>>16, img.height>>24};
  uint8 w[4] = {img.width,  img.width >>8, img.width >>16, img.width >>24};
  if (fwrite(h, sizeof(h), 1, f) < 1) panic("Cannot write image height", errno);
  if (fwrite(w, sizeof(w), 1, f) < 1) panic("Cannot write image width", errno);

  // Write pixel data
  int img_size = img.height * img.width * img.channels;
  if (fwrite(img.data, sizeof(uint8), img_size, f) < img_size) {
    panic("Cannot write image data", errno);
  }

  // Clean up and return
  fclose(f);
}


